lib/scripts/nc.py

```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env python
+
+import os
+import socket
+import sys
+import threading
+
+
+class Sender(threading.Thread):
+    def __init__(self, sock):
+        super(Sender, self).__init__()
+        self.sock = sock
+
+    def run(self):
+        line = sys.stdin.readline()
+        while line:
+            try:
+                sock.sendall(line)
+            except Exception as e:
+                try:
+                    sock.shutdown(socket.SHUT_RDWR)
+                    sock.close()
+                except Exception:
+                    pass
+                exit(1)
+            line = sys.stdin.readline()
+
+
+try:
+    sock = socket.create_connection(("localhost", sys.argv[1]), timeout=None)
+except Exception as e:
+    exit(1)
+
+sender = Sender(sock)
+sender.daemon = True
+sender.start()
+
+# HACK: Try to read from the socket, which will block as long as the connection is OK since
+# the agent isn't sending anything on the connection. As soon as this returns, it means the
+# connection was closed on the other end (agent times out the connection if it does not receive
+# logs during 1 minute). This is to prevent losing logs because after such a disconnect, the first
+# message sent on the socket will succeed, even though it cannot possibly have been received.
```

Comment:
Instead of doing that, could you handle the error in `sender` and retry to connect? That would avoid exiting on error and just retrying instead?

Reply:
That's the point, I can't do that because that would mean losing a line since it would appear to succeed on the first try after the remote connection has been shutdown by the agent

Reply:
AFAIU `send` should fail if the connection is closed? At that point in python you still have the line data and you can retry.

Reply:
That's what one would assume I agree 😅, but for some reason it's not the case. Believe me that's been bugging me a long while.

Try this for yourself 
```python
import socket

input("Start a netcat in a terminal and press enter: `nc -l localhost 15000`")
sock = socket.create_connection(("localhost", 15000), timeout=None)
sock.sendall(b"This line arrives")
input("Interrupt netcat and press enter to continue")
sock.sendall(b"This line gets lost in limbo")
print("Script still running")
sock.sendall(b"This finally errors out, but the previous line was lost")
print("Script has errored out")
```

Reply:
Ah right we'd need to use a mechanism like select.